src/Connection.ts

```diff
@@ -1,1488 +1,1884 @@
+			this.connected = true;
+
+			if (this.waitForRestart) {
+				window.location.reload();
+			} else {
+				this._subscribe(true);
+				this.onConnectionHandlers.forEach((cb) => cb(true));
+			}
+		});
+
+		this._socket.on("disconnect", () => {
+			this.connected = false;
+			this.subscribed = false;
+			this.props.onProgress?.(PROGRESS.CONNECTING);
+			this.onConnectionHandlers.forEach((cb) => cb(false));
+		});
+
+		this._socket.on("reauthenticate", () => this.authenticate());
+
+		this._socket.on("log", (message) => {
+			this.props.onLog?.(message);
+			this.onLogHandlers.forEach((cb) => cb(message));
+		});
+
+		this._socket.on("error", (err: any) => {
+			let _err: string;
+
+			if (err == undefined) {
+				_err = "";
+			} else if (typeof err.toString !== "function") {
+				_err = err.toString();
+			} else {
+				_err = JSON.stringify(err);
+				console.error(`Received strange error: ${_err}`);
+			}
+
+			if (_err.includes("User not authorized")) {
+				this.authenticate();
+			} else {
+				window.alert(`Socket Error: ${err}`);
+			}
+		});
+
+		this._socket.on("connect_error", (err: any) =>
+			console.error(`Connect error: ${err}`),
+		);
+
+		this._socket.on("permissionError", (err) =>
+			this.onError({
+				message: "no permission",
+				operation: err.operation,
+				type: err.type,
+				id: err.id || "",
+			}),
+		);
+
+		this._socket.on("objectChange", (id, obj) => {
+			setTimeout(() => this.objectChange(id, obj), 0);
+		});
+
+		this._socket.on("stateChange", (id, state) => {
+			setTimeout(() => this.stateChange(id, state), 0);
+		});
+
+		this._socket.on("cmdStdout", (id, text) => {
+			this.onCmdStdoutHandler?.(id, text);
+		});
+
+		this._socket.on("cmdStderr", (id, text) => {
+			this.onCmdStderrHandler?.(id, text);
+		});
+
+		this._socket.on("cmdExit", (id, exitCode) => {
+			this.onCmdExitHandler?.(id, exitCode);
+		});
+	}
+
+	/**
+	 * Called internally.
+	 * @param isOk
+	 * @param isSecure
+	 */
+	private onPreConnect(isOk: boolean, isSecure: boolean) {
+		if (this._authTimer) {
+			clearTimeout(this._authTimer);
+			this._authTimer = null;
+		}
+
+		this.connected = true;
+		this.isSecure = isSecure;
+
+		if (this.waitForRestart) {
+			window.location.reload();
+		} else {
+			if (this.firstConnect) {
+				this.loadData();
+			} else {
+				this.props.onProgress?.(PROGRESS.READY);
+			}
+
+			this._subscribe(true);
+			this.onConnectionHandlers.forEach((cb) => cb(true));
+		}
+
+		this._waitForFirstConnectionPromise.resolve();
+	}
+
+	/**
+	 * Checks if the socket is connected.
+	 * @returns {boolean} true if connected.
+	 */
+	isConnected(): boolean {
+		return this.connected;
+	}
+
+	/**
+	 * Checks if the socket is connected.
+	 * @returns {Promise<void>} Promise resolves if once connected.
+	 */
+	waitForFirstConnection(): Promise<void> {
+		return this._waitForFirstConnectionPromise;
+	}
+
+	/**
+	 * Called internally.
+	 */
+	private async getUserPermissions(): Promise<any> {
```

Comment:
Probably for later: this shouldn't be `any`.

Reply:
It definitely shouldn't